packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,304 @@
+            return false;
+          } else { // we matched, move to next character in input for comparison
+            match = true;
+          }
+        }
+      }
+    }
+    return true;
+  }, [collator, lowercaseValue]);
+
+  useEffect(() => {
+    if (onFilter) {
+      onFilter(value);
+    }
+  // Having onFilter in the dep array seems to break it
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [value]);
+
+  selectState.collection = useMemo(() => {
+    if (itemsControlled || value === '') {
+      return selectState.collection;
+    }
+    return new FilteredCollection(selectState.collection, defaultFilterFn);
+  }, [selectState.collection, value, itemsControlled, defaultFilterFn]);
+ 
+  useEffect(() => {
+    // Close the menu if it was open but there aren't any items to display
+    if (menuIsOpen && selectState.collection.size === 0) {
+      selectState.close();
+    }
+
+    // Only open the menu if there are items to display and the combobox is focused
+    // Note: doesn't affect controlled isOpen or defaultOpen
+    if (isFocused && menuIsOpen && selectState.collection.size > 0) {
+      selectState.open();
+    }
+
+    // Close the menu if it is supposed to be closed
+    if (!menuIsOpen) {
+      selectState.close();
+    }
+
+    // Maybe change dep array back to selectState? or make it selectState.close?
+  }, [menuIsOpen, selectState.collection, selectState, isFocused]);
+
+  let open = () => {
+    setMenuIsOpen(true);
+  };
+
+  let close = () => {
+    setMenuIsOpen(false);
+  };
+
+  let toggle = (focusStrategy = null) => {
+    setMenuIsOpen(state => !state);
+    selectState.setFocusStrategy(focusStrategy);
+  };
+
+  // Moved from aria to stately cuz it feels more like stately
+  useEffect(() => {
+    // Perhaps replace the below with state.selectedItem?
+    let selectedItem = selectState.selectedKey ? selectState.collection.getItem(selectState.selectedKey) : null;
+    if (selectedItem) {
+      let itemText = selectedItem.textValue || selectedItem.rendered as string; // how should we handle this? rendered is typed as an object
+
+      // Throw error if controlled inputValue and controlled selectedKey don't match
+      if (inputValue && selectedKey && (inputValue !== itemText)) {
+        throw new Error('Mismatch between selected item and inputValue!');
+      }
+
+      // Update textfield value if new item is selected
+      // Only do this if not controlled?
+      if (itemText !== value && !(inputValue)) {
+        setValue(itemText);
+      }
+    } else {
+      if (inputValue) {
+        // TODO find item that has matching text and set as selectedKey
+        // If none found, make invalid?
+      }
+    }
+  // Double check this dependency array (does it need value,setValue, selectState.collection)
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [selectState.selectedKey, inputValue, selectedKey]);
```

Comment:
I think this `useEffect` could go away if you implemented the suggestion above to store the input value and selected key separately, and return a computed value below (e.g. `value: inputValue || selectState.selectedKey`. That way you wouldn't need to synchronize the state here.